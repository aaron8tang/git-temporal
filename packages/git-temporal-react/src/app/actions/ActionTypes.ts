export enum ActionTypes {
  REQUEST_COMMITS = 'REQUEST_COMMITS',
  RECEIVE_COMMITS = 'RECEIVE_COMMITS',
  SELECT_PATH = 'SELECT_PATH',
  INVALIDATE_PATH = 'INVALIDATE_PATH',
  HIGHLIGHT_COMMIT = 'HIGHLIGHT_COMMIT',
  VIEW_COMMITS = 'VIEW_COMMITS',
  VIEW_FILES = 'VIEW_FILES',
  ADD_AUTHOR_FILTER = 'ADD_AUTHOR_FILTER',
  REMOVE_AUTHOR_FILTER = 'REMOVE_AUTHOR_FILTER',
  REMOVE_ALL_AUTHOR_FILTERS = 'REMOVE_ALL_AUTHOR_FILTERS',
}
