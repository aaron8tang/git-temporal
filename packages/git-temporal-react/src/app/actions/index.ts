import { ActionTypes } from './ActionTypes';

export const selectPath = path => (dispatch, _getState) => {
  // if this comes from a rename, follow the most current name
  const actualPath = path.replace(/\{(.*)\s=>\s(.*)\}/g, '$2');

  dispatch(fetchCommitsIfNeeded(actualPath));
  return { selectedPath: actualPath, type: ActionTypes.SELECT_PATH };
};

export const invalidatePath = path => ({
  selectedPath: path,
  type: ActionTypes.INVALIDATE_PATH,
});

export const addAuthorFilter = authorName => ({
  authorName,
  type: ActionTypes.ADD_AUTHOR_FILTER,
});

export const removeAuthorFilter = authorName => ({
  authorName,
  type: ActionTypes.REMOVE_AUTHOR_FILTER,
});

export const removeAllAuthorFilters = () => ({
  type: ActionTypes.REMOVE_ALL_AUTHOR_FILTERS,
});

export const highlightCommit = commitId => ({
  commitId,
  type: ActionTypes.HIGHLIGHT_COMMIT,
});

export const viewCommits = () => ({
  type: ActionTypes.VIEW_COMMITS,
});

export const viewFiles = () => ({
  type: ActionTypes.VIEW_FILES,
});

export const requestCommits = path => ({
  selectedPath: path,
  type: ActionTypes.REQUEST_COMMITS,
});

export const receiveCommits = (path, json) => ({
  selectedPath: path,
  commits: json.commits,
  type: ActionTypes.RECEIVE_COMMITS,
});

const fetchCommits = path => dispatch => {
  dispatch(requestCommits(path));
  const pathParam = path && path.trim().length > 0 ? `?path=${path}` : '';
  // TODO : replace this with serviceBaseUrl when it is in
  return fetch(`http://localhost:11966/git-temporal/history${pathParam}`)
    .then(response => response.json())
    .then(json => dispatch(receiveCommits(path, json)));
};

const shouldFetchCommits = (state, path) => {
  if (state.isFetching) {
    return false;
  }
  return (
    state.didInvalidate ||
    !state.commits ||
    state.commits.length <= 0 ||
    state.selectedPath !== path
  );
};

export const fetchCommitsIfNeeded = path => (dispatch, getState) => {
  if (shouldFetchCommits(getState(), path)) {
    return dispatch(fetchCommits(path));
  }
};
