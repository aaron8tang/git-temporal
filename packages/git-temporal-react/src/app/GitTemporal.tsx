import React, { Component } from 'react';
import { connect } from 'react-redux';
import { debug } from '@git-temporal/logger';

import { GitTemporalProps, DispatchProps, StateProps } from 'app/interfaces';
import { selectPath } from 'app/actions';
import { handleVscodeMessages } from 'app/actions/vscode';
import { getGitTemporalContainerState } from 'app/selectors';
import { style } from 'app/styles';

import Header from 'app/containers/Header';
import Stats from 'app/containers/Stats';
import Authors from 'app/containers/Authors';
import Files from 'app/containers/Files';
import Commits from 'app/containers/Commits';
import Timeplot from 'app/containers/Timeplot';
import DifferenceViewer from 'app/containers/DifferenceViewer';

import { SpinnerContainer } from 'app/components/SpinnerContainer';
import { HorizontalScroller } from 'app/components/HorizontalScroller';
import { TransitionVisible } from 'app/components/TransitionVisible';

const outerScrollStyle = {
  _extends: 'flexRows',
  position: 'relative',
  minHeight: '100%',
  height: 'initial',
};

const innerScrollStyle = {
  position: 'relative',
  display: 'flex',
  flexDirection: 'row',
  flexGrow: 1,
};

const scrollerIconStyle = {
  transform: 'scaleY(8)',
  top: '35%',
};

const transitionStyle = {
  _extends: 'flexColumns',
  flexGrow: 1,
};

export class GitTemporal extends Component<
  GitTemporalProps & DispatchProps & StateProps
> {
  componentDidMount() {
    const { path, dispatch } = this.props;
    dispatch(selectPath(path));
    handleVscodeMessages(dispatch);
  }

  componentWillUnmount() {
    debug('unmounting GitTemporal');
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.path !== this.props.path) {
      const { dispatch, path } = nextProps;
      dispatch(selectPath(path));
    }
  }

  render() {
    const { isFetching, viewCommitsOrFiles } = this.props;
    debug('rendering GitTemporal');

    return (
      <div style={style('page')}>
        <SpinnerContainer isSpinning={isFetching}>
          <div style={style('flexColumns', { height: '100%' })}>
            <Header />
            <TransitionVisible
              isVisible={!isFetching}
              style={style(transitionStyle)}
            >
              <Stats />
              <div style={{ display: 'flex', flexGrow: 1 }}>
                <HorizontalScroller
                  style={style(outerScrollStyle)}
                  innerStyle={style(innerScrollStyle)}
                  iconStyle={style(scrollerIconStyle)}
                >
                  <Authors />
                  {viewCommitsOrFiles === 'files' ? <Files /> : <Commits />}
                  <DifferenceViewer />
                </HorizontalScroller>
              </div>
              <Timeplot />
            </TransitionVisible>
          </div>
        </SpinnerContainer>
      </div>
    );
  }
}

export default connect(getGitTemporalContainerState)(GitTemporal);
