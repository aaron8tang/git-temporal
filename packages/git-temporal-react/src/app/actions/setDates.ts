import { setStartDate, setEndDate } from 'app/actions';

export function setDates(dispatch, startDate, endDate, preferEndDate, date) {
  const epochDate = Math.floor(date / 1000);
  if (!startDate && !endDate) {
    dispatch(setStartDate(epochDate));
  } else if (startDate) {
    if (preferEndDate) {
      if (epochDate < startDate) {
        dispatch(setEndDate(startDate));
        dispatch(setStartDate(epochDate));
      } else {
        dispatch(setEndDate(epochDate));
      }
    } else {
      if (endDate && epochDate > endDate) {
        dispatch(setStartDate(endDate));
        dispatch(setEndDate(epochDate));
      } else {
        dispatch(setStartDate(epochDate));
      }
    }
  }
}
