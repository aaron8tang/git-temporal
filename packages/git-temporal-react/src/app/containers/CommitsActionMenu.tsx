import React, { Component } from 'react';
import { connect } from 'react-redux';
import { style } from 'app/styles';

import { DispatchProps } from 'app/interfaces';
import { getCommitsActionMenuState } from 'app/selectors';
import { RadioMenuItem } from 'app/components/RadioMenuItem';
import { ActionMenu } from 'app/components/ActionMenu';

import { setCommitsContainerSort } from 'app/actions';
import { CommitsContainerSorts } from 'app/actions/ActionTypes';

interface CommitsActionMenuProps {
  commitsContainerSort?: CommitsContainerSorts;
}

const containerStyle = {
  _extends: 'normalText',
  position: 'absolute',
  right: 10,
};

export class CommitsActionMenu extends Component<
  CommitsActionMenuProps & DispatchProps
> {
  render() {
    const { commitsContainerSort } = this.props;
    return (
      <ActionMenu style={style(containerStyle)}>
        <div style={style('h5Text')}>Sort by</div>
        <RadioMenuItem
          testId="sortTime"
          onClick={() => {
            this.sortCommits(CommitsContainerSorts.TIME);
          }}
          isSelected={commitsContainerSort === CommitsContainerSorts.TIME}
        >
          by time
        </RadioMenuItem>
        <RadioMenuItem
          testId="sortLines"
          onClick={() => {
            this.sortCommits(CommitsContainerSorts.LINES);
          }}
          isSelected={commitsContainerSort === CommitsContainerSorts.LINES}
        >
          by lines changed
        </RadioMenuItem>
      </ActionMenu>
    );
  }

  sortCommits(newCommitsContainerSort: CommitsContainerSorts) {
    this.props.dispatch(setCommitsContainerSort(newCommitsContainerSort));
  }
}

export default connect(getCommitsActionMenuState)(CommitsActionMenu);
