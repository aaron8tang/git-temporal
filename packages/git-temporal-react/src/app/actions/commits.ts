import { ActionTypes } from 'app/actions/ActionTypes';
import { isVscode, vscode } from 'app/actions/vscode';
import { debug } from '@git-temporal/logger';

export const requestCommits = path => ({
  selectedPath: path,
  type: ActionTypes.REQUEST_COMMITS,
});

export const receiveCommits = (path, json) => ({
  selectedPath: path,
  commits: json.commits,
  isFileSelected: json.isFile,
  type: ActionTypes.RECEIVE_COMMITS,
});

const fetchCommits = path => dispatch => {
  dispatch(requestCommits(path));
  if (isVscode) {
    debug('sending history request to vscode ', path);
    // see actions/vscode.ts for response handling that comes as a window event
    vscode.postMessage({ path, command: 'history' });
  } else {
    const pathParam = path && path.trim().length > 0 ? `?path=${path}` : '';
    // TODO : replace this with serviceBaseUrl when it is in
    fetch(`http://localhost:11966/git-temporal/history${pathParam}`)
      .then(response => response.json())
      .then(json => dispatch(receiveCommits(path, json)));
  }
};

const shouldFetchCommits = (state, path) => {
  if (state.isFetching) {
    return false;
  }
  return (
    !state.commits || state.commits.length <= 0 || state.selectedPath !== path
  );
};

export const fetchCommitsIfNeeded = path => (dispatch, getState) => {
  if (shouldFetchCommits(getState(), path)) {
    return dispatch(fetchCommits(path));
  }
};
