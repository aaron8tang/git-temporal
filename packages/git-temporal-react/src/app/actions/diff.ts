import { ICommit } from 'app/interfaces';
import { dateFilteredCommits } from 'app/selectors/dates';
import { ActionTypes } from 'app/actions/ActionTypes';
import { isVscode, vscode } from 'app/actions/vscode';
import { debug } from '@git-temporal/logger';
import { start } from 'repl';

const requestDiff = (path: string) => ({
  selectedPath: path,
  type: ActionTypes.REQUEST_DIFF,
});

export const receiveDiff = (path: string, diff: any) => ({
  diff,
  selectedPath: path,
  type: ActionTypes.RECEIVE_DIFF,
});

export const setDiffStartCommit = (commitId: string) => ({
  commitId,
  type: ActionTypes.SET_DIFF_START_COMMIT,
});

export const setDiffEndCommit = (commitId: string) => ({
  commitId,
  type: ActionTypes.SET_DIFF_END_COMMIT,
});

export const fetchDiff = (
  path: string,
  commits: ICommit[],
  startDate: number,
  endDate: number
) => (dispatch: any): void => {
  dispatch(requestDiff(path));
  const filteredCommits = dateFilteredCommits(commits, startDate, endDate);
  if (!filteredCommits || filteredCommits.length === 0) {
    return;
  }
  // commits are in descending order
  const rightCommit = filteredCommits[0];
  const leftCommit = filteredCommits.slice(-1)[0];

  if (isVscode) {
    debug('sending diff request to vscode ', path, leftCommit, rightCommit);
    // see actions/vscode.ts for response handling that comes as a window event
    vscode.postMessage({
      path,
      leftCommit: leftCommit.id,
      rightCommit: rightCommit.id,
      command: 'diff',
    });
  } else {
    const pathParam =
      path && path.trim().length > 0 ? `?path=${path}` : '?path=.';
    const leftCommitParam = leftCommit ? `&leftCommit=${leftCommit.id}` : '';
    const rightCommitParam = rightCommit
      ? `&rightCommit=${rightCommit.id}`
      : '';

    // TODO : replace this with serviceBaseUrl when it is in
    const url = `http://localhost:11966/git-temporal/diff${pathParam}${leftCommitParam}${rightCommitParam}`;
    fetch(url)
      .then(response => response.json())
      .then(diff => dispatch(receiveDiff(path, diff)));
  }
};
