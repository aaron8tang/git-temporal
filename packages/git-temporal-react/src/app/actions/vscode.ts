import { receiveCommits, receiveCommitRange } from 'app/actions/commits';
import { receiveDiff } from 'app/actions/diff';
import { debug } from 'app/utilities/logger';

// @ts-ignore
export const isVscode = window && window.IS_VSCODE_WEBVIEW;
// @ts-ignore
export const vscode = isVscode ? acquireVsCodeApi() : null;

if (isVscode) {
  debug('running in VSCode.');
}

export function handleVscodeMessages(dispatch) {
  // @ts-ignore
  if (!window || !window.IS_VSCODE_WEBVIEW) {
    return;
  }
  window.addEventListener('message', event => {
    const { data } = event;
    debug(`actions/vscode received window message`, {
      ...data,
      // don't dump full commits or file contents to log
      commits: data.commits ? `object[${data.commits.length}]` : undefined,
      leftFileContents: redactArray(data.leftFileContents),
      rightFileContents: redactArray(data.rightFileContents),
      modifiedFiles: redactArray(data.modifiedFiles),
    });

    switch (data.type) {
      case 'commitRange':
        dispatch(receiveCommitRange(data.path, data));
        break;
      case 'history':
        dispatch(receiveCommits(data.path, data));
        break;
      case 'diff':
        dispatch(receiveDiff(data.path, data));
        break;
    }
  });
}

function redactArray(array) {
  return array ? `[..] length=${array.length}` : undefined;
}
