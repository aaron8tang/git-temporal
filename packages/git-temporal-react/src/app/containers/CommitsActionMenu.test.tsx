import React from 'react';
import * as actions from 'app/actions';
import CommitsActionMenu from './CommitsActionMenu';

import { fireEvent } from '@testing-library/react';
import { mountConnected } from 'testHelpers';

describe('containers/CommitsActionMenu', () => {
  describe('when rendered with mock redux store', () => {
    test('it should match snapshot', () => {
      const { wrapper } = mountConnected(<CommitsActionMenu />);
      expect(wrapper.baseElement).toMatchSnapshot();
    });

    describe('...and then clicking menu items', () => {
      const menusToTest = [
        { testId: 'sortTime', actionFn: actions.setCommitsContainerSort },
        { testId: 'sortLines', actionFn: actions.setCommitsContainerSort },
      ];
      beforeEach(() => {
        jest.clearAllMocks();
      });
      for (const menuTest of menusToTest) {
        test(`it should call actions in response to ${
          menuTest.testId
        } item clicks`, () => {
          const { wrapper, store } = mountConnected(<CommitsActionMenu />);
          const menuItem = wrapper.getByTestId(menuTest.testId);
          fireEvent.click(menuItem);
          expect(store.dispatch).toHaveBeenCalledTimes(1);
        });
      }
    });
  });
});
