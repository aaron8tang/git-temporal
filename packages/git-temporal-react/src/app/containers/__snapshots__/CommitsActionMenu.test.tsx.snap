// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/CommitsActionMenu when rendered with basic props it should match snapshot 1`] = `
<ActionMenu
  isMenuOpen={false}
  onMenuToggle={[Function]}
  style={
    Object {
      "color": "#333333",
      "fontSize": 13,
      "fontWeight": "normal",
      "lineHeight": 1.35,
      "position": "absolute",
      "right": 10,
    }
  }
>
  <div
    style={
      Object {
        "color": "#333333",
        "fontSize": 14,
        "fontWeight": "normal",
        "lineHeight": 1.35,
        "marginBottom": 10,
        "marginRight": 10,
      }
    }
  >
    Sort by
  </div>
  <RadioMenuItem
    isSelected={true}
    onClick={[Function]}
    testId="sortTime"
  >
    by time
  </RadioMenuItem>
  <RadioMenuItem
    isSelected={false}
    onClick={[Function]}
    testId="sortLines"
  >
    by lines changed
  </RadioMenuItem>
</ActionMenu>
`;
